README.md
usersd
======

usersd is a DBus activatable service that permits to manage and create
users and groups.

Design
------

Every user is exported as a separate DBus object, at the path:

	/org/semplicelinux/usersd/user/UID

where UID is the user's UID.

Properties (Full Name, Home directory, Address, etc) are exported through
DBus' standard Properties interface, but they aren't introspected due to
technical implications.

Properties are writeable, and they are syncronized automatically to /etc/passwd.

Likewise, every group is exported as a separate DBus object, too:

	/org/semplicelinux/usersd/group/GID

where GID is the group's GID.

Group members are exported as a property.

Security
--------

usersd runs as root, but requires Polkit authentication from the caller user
before actually doing things.

An exception is made for the caller user's object. Caller users can change
their properties and passwords as much as they want, but they can't delete
themselves.

To avoid sending password hashes through the system bus, the user creation and
password change methods are only available via a service-side GUI.


debian/changelog
usersd (0.90.2-1) unstable; urgency=medium

  * Use full path when calling deluser.

 -- Eugenio Paolantonio (g7) <me@medesimo.eu>  Thu, 02 Apr 2015 17:52:16 +0200

usersd (0.90.1-1) unstable; urgency=medium

  * Use full path when calling useradd.

 -- Eugenio Paolantonio (g7) <me@medesimo.eu>  Thu, 02 Apr 2015 17:49:27 +0200

usersd (0.90-1) unstable; urgency=low

  * Initial release.

 -- Eugenio Paolantonio (g7) <me@medesimo.eu>  Thu, 02 Apr 2015 14:59:23 +0200


usersd/ui.py
# -*- coding: utf-8 -*-
#
# usersd - user management daemon
# Copyright (C) 2014  Eugenio "g7" Paolantonio
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
# Authors:
#    Eugenio "g7" Paolantonio <me@medesimo.eu>
#

from gi.repository import Gtk, GObject

import quickstart

class UI:
	
	"""
	The usersd user interface
	"""
	
	events = {
	
	}
	
	map_to = None
	error_message = None
	error_revealer = None
	
	def __getattr__(self, key):
		"""
		Proxy to the internal change_password_dialog.
		"""
		
		if self.map_to:
			return getattr(self.objects[self.map_to], key)
		else:
			# If we are here we don't have any other attributes to return
			raise AttributeError("UI object has no attribute '%s'" % key)
	
	def show_error(self, message):
		"""
		Shows an error message.
		"""
		
		if self.error_message:
			self.objects[self.error_message].set_text(message)
		if self.error_revealer:
			GObject.idle_add(self.objects[self.error_revealer].set_reveal_child, True)
	
	def hide_error(self):
		"""
		Hides the error message.
		"""
		
		if self.error_revealer:
			GObject.idle_add(self.objects[self.error_revealer].set_reveal_child, False)
	
	def __init__(self):
		"""
		Initializes the class.
		"""
		
		quickstart.events.connect(self)

@quickstart.builder.from_file("./usersd/usersd.glade")
class ChangePasswordDialog(UI):
	
	map_to = "change_password_dialog"
	error_message = "error_message"
	error_revealer = "error_revealer"
	
	def __init__(self, locked=False):
		"""
		Initializes the class.
		"""
		
		super().__init__()

		self.locked = locked
		
		# Add buttons
		self.objects.change_password_dialog.add_buttons(
			_("_Cancel"),
			Gtk.ResponseType.CANCEL,
			_("Ch_ange"),
			Gtk.ResponseType.OK
		)
		self.objects.change_password_dialog.set_default_response(Gtk.ResponseType.OK)
		
		# Remove 'Old password' requirement if locked
		if self.locked:
			self.objects.old_password_label.hide()
			self.objects.old_password.hide()

@quickstart.builder.from_file("./usersd/usersd.glade")
class AddUserDialog(UI):
	
	map_to = "add_user_dialog"
	error_message = "add_error_message"
	error_revealer = "add_error_revealer"
	
	def __init__(self):
		"""
		Initializes the class.
		"""
		
		super().__init__()
		
		# Add buttons
		self.objects.add_user_dialog.add_buttons(
			_("_Cancel"),
			Gtk.ResponseType.CANCEL,
			_("_Add"),
			Gtk.ResponseType.OK
		)
		self.objects.add_user_dialog.set_default_response(Gtk.ResponseType.OK)


usersd/group.py
# -*- coding: utf-8 -*-
#
# usersd - user management daemon
# Copyright (C) 2014  Eugenio "g7" Paolantonio
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
# Authors:
#    Eugenio "g7" Paolantonio <me@medesimo.eu>
#

import usersd.objects
import subprocess

class Group(usersd.objects.BaseObject):
	"""
	The Group object
	"""
	
	interface_name = "org.semplicelinux.usersd.group"
	export_properties = [
		"group",
		"gid",
		"members",
	]
	polkit_policy = "org.semplicelinux.usersd.modify-group"
	
	def __init__(self, service, bus_name, group_entry):
		"""
		Initializes the object.
		"""
		
		self.service = service
		
		self.group, password, gid, members = group_entry.split(":")
		
		self.gid = int(gid)
		self.members = members.split(",")
		
		self.set_privileges = []
		
		self.path = "/org/semplicelinux/usersd/group/%s" % gid
		super().__init__(bus_name)
	
	def refresh_members_from_group_entry(self, group_entry):
		"""
		Refreshes the members list from a group_entry line.
		"""
		
		self.members = group_entry.split(":")[-1].split(",")
	
	def store_property(self, name, value):
		"""
		Stores the modified property in the /etc/passwd file.
		"""

		if name.lower() == "members":
			# Use gpasswd
			to_add = [user for user in value if not user in self.members]
			to_remove = [user for user in self.members if not user in value]
						
			if to_add:
				for user in to_add:
					subprocess.call((
						"gpasswd",
						self.group,
						"-a",
						user
					))
			
			if to_remove:
				for user in to_remove:
					subprocess.call((
						"gpasswd",
						self.group,
						"-d",
						user
					))
		else:
			# Not supported for now
			return

		# Set value
		setattr(self, name[0].lower() + name[1:], value)



usersd/common.py
# -*- coding: utf-8 -*-
#
# usersd - user management daemon
# Copyright (C) 2014  Eugenio "g7" Paolantonio
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
# Authors:
#    Eugenio "g7" Paolantonio <me@medesimo.eu>
#

import os

import dbus

import importlib

from gi.repository import GLib, Polkit

authority = Polkit.Authority.get_sync()

class ModuleProxy:
	"""
	Transparent proxy to a python module.
	
	This proxy loads the specified module only when another object tries
	to access it, and not on load.
	"""
	
	_real = None
	
	def _initialize(self, display, user, dct):
		"""
		Initializes core variables to access the current user's display.
		"""
		
		os.environ["XAUTHORITY"] = os.path.join(dct[user].home, ".Xauthority")
		os.environ["DISPLAY"] = display
	
	def __init__(self, module):
		"""
		Initializes the class.
		"""
		
		self.module = module
	
	def __getattr__(self, attr):
		"""
		Where the magic happens.
		"""
		
		if not self._real:
			self._real = importlib.import_module(self.module)
		
		return getattr(self._real, attr)

class LoopWithTimeout():
	"""
	A LoopWithTimeout is a GLibMainLoop that supports timeouts.
	
	We don't directly subclass GLib.MainLoop because we can't.
	"""
	
	def __init__(self, timeout_length):
		"""
		Initializes the loop.
		"""
		
		self.loop = GLib.MainLoop()
		
		self.timeout_length = timeout_length
		
		self.add_timeout()

	def on_timeout_elapsed(self):
		"""
		Fired when the timeout elapsed.
		"""
		
		self.quit()
		
		return False
	
	def remove_timeout(self):
		"""
		Removes the timeout.
		"""

		if self.timeout > 0:
			# Timeout already present, cancel it
			GLib.source_remove(self.timeout)
			self.timeout = 0
	
	def add_timeout(self):
		"""
		Timeout.
		"""
		
		self.timeout = GLib.timeout_add_seconds(self.timeout_length, self.on_timeout_elapsed)
	
	def __getattr__(self, name):
		"""
		Proxy to the self.loop object.
		"""
		
		return getattr(self.loop, name)

def get_user(sender):
	"""
	Returns the user name of the given sender.
	"""
	
	return dbus.Interface(
		dbus.SystemBus().get_object(
			"org.freedesktop.DBus",
			"/org/freedesktop/DBus"
		),
		"org.freedesktop.DBus"
	).GetConnectionUnixUser(sender)

def is_authorized(sender, connection, privilege, user_interaction=True):
	"""
	Checks if the sender has the given privilege.
	
	Returns True if yes, False if not.
	"""
	
	if not user_interaction:
		flags = Polkit.CheckAuthorizationFlags.NONE
	else:
		flags = Polkit.CheckAuthorizationFlags.ALLOW_USER_INTERACTION
	
	# Get PID
	pid = dbus.Interface(
		dbus.SystemBus().get_object(
			"org.freedesktop.DBus",
			"/org/freedesktop/DBus"
		),
		"org.freedesktop.DBus"
	).GetConnectionUnixProcessID(sender)
		
	try:
		result = authority.check_authorization_sync(
			Polkit.UnixProcess.new(pid),
			privilege,
			None,
			flags,
			None
		)
	except:
		return False
	
	return result.get_is_authorized()

MainLoop = LoopWithTimeout(5 * 60)

usersd_ui = ModuleProxy("usersd.ui")
Gtk = ModuleProxy("gi.repository.Gtk")


usersd-service.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
#
# usersd - user management daemon
# Copyright (C) 2014  Eugenio "g7" Paolantonio
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
# Authors:
#    Eugenio "g7" Paolantonio <me@medesimo.eu>
#

import os

import dbus

from usersd.common import MainLoop, is_authorized

import usersd.objects
import usersd.user
import usersd.group

import quickstart.translations

from dbus.mainloop.glib import DBusGMainLoop

if os.path.islink(__file__):
	# If we are a link, everything is a WTF...
	USERSD_DIR = os.path.dirname(os.path.normpath(os.path.join(os.path.dirname(__file__), os.readlink(__file__))))
else:
	USERSD_DIR = os.path.dirname(__file__)


# While the following is not ideal, is currently needed to make sure
# we are actually on the main vera-control-center directory.
# The main executable (this) and all modules do not use absolute paths
# to load the glade UI files, so we need to be on the main directory
# otherwise they will crash.
# This should be probably addressed directly in quickstart.builder but,
# for now, this chdir call will do the job.
os.chdir(USERSD_DIR)

# Parse default locale from /etc/default/locale
try:
	with open("/etc/default/locale", "r") as f:
		os.environ["LANG"] = f.readline().strip().split("=")[-1]
except:
	pass

TRANSLATION = quickstart.translations.Translation("usersd")
TRANSLATION.load()
TRANSLATION.install()
TRANSLATION.bind_also_locale()

class Usersd(usersd.objects.BaseObject):
	"""
	The main object.
	"""
	
	path = "/org/semplicelinux/usersd"
	
	@dbus.service.signal(
		"org.semplicelinux.usersd.user"
	)
	def UserListChanged(self):
		"""
		Signal emitted when the user list has been changed.
		"""
		
		pass
	
	def __init__(self):
		"""
		Initializes the object.
		"""
		
		self.bus_name = dbus.service.BusName(
			"org.semplicelinux.usersd",
			bus=dbus.SystemBus()
		)
		
		super().__init__(self.bus_name)
		
		self._users = {}
		self._groups = {}
		self._generate_users(refresh_groups=False)
		self._generate_groups()
	
	def _generate_users(self, refresh_groups=True):
		"""
		Generates a user object for every user in /etc/passwd.
		"""
				
		with open("/etc/passwd", "r") as f:
			for user in f:
				name = user.split(":")[0]
				if not name in self._users:
					self._users[name] = usersd.user.User(
						self,
						self.bus_name,
						user.strip()
					)
		
		# Refresh groups if asked to
		if refresh_groups:
			self._generate_groups()
		
		# Emit signal
		self.UserListChanged()
	
	def _generate_groups(self, refresh=False):
		"""
		Generates a group object for every group in /etc/group.
		"""
		
		with open("/etc/group", "r") as f:
			for group in f:
				name = group.split(":")[0]
				if not name in self._groups:
					self._groups[name] = usersd.group.Group(
						self,
						self.bus_name,
						group.strip()
					)
				elif refresh:
					self._groups[name].refresh_members_from_group_entry(group.strip())
	
	def remove_from_user_list(self, user):
		"""
		Removes the given username from the users list.
		"""
		
		if user in self._users:
			del self._users[user]

		# Refresh groups
		self._generate_groups(refresh=True)
		
		# Emit signal
		self.UserListChanged()

	@usersd.objects.BaseObject.outside_timeout(
		"org.semplicelinux.usersd.group",
		out_signature="a{i(s)}",
		sender_keyword="sender",
		connection_keyword="connection"
	)
	def GetGroups(self, sender, connection):
		"""
		This method returns a dictionary containing every group's GID as keys,
		and the groupname as values.
		"""
		
		result = {}
					
		for group, obj in self._groups.items():
			result[obj.gid] = (group,)
		
		return result

	@usersd.objects.BaseObject.outside_timeout(
		"org.semplicelinux.usersd.group",
		in_signature="s",
		out_signature="as",
		sender_keyword="sender",
		connection_keyword="connection"
	)
	def GetGroupsForUser(self, user, sender, connection):
		"""
		This method returns an array containing every group the given
		user is in.
		"""
		
		result = []
					
		for group, obj in self._groups.items():
			if user in obj.members:
				result.append(group)
		
		return result

	@usersd.objects.BaseObject.outside_timeout(
		"org.semplicelinux.usersd.group",
		in_signature="s",
		out_signature="s",
		sender_keyword="sender",
		connection_keyword="connection"
	)
	def LookupGroup(self, group, sender, connection):
		"""
		This method returns the object path for the given group.
		"""
		
		if group in self._groups: return self._groups[group].path
		
		return None
	
	# NOTE: The following method needs to be properly security-audited
	# before we can export it to DBus.
	# For now is only meant to be used internally when creating an user.
	# You can take advantage of that using the ShowUserCreationUI method.
	#
	#@usersd.objects.BaseObject.outside_timeout(
	#	"org.semplicelinux.usersd.user",
	#	in_signature="sas",
	#	sender_keyword="sender",
	#	connection_keyword="connection"
	#)
	def AddGroupsToUser(self, user, groups, sender=None, connection=None):
		"""
		Adds the given user to every group specified in the specfied
		groups list.
		"""
		
		if sender and connection and not is_authorized(
			sender,
			connection,
			"org.semplicelinux.usersd.add-user",
			True # user interaction
		):
			raise Exception("Not authorized")
		
		for group in groups:
			if not group in self._groups:
				continue
			
			members = self._groups[group].members
			if not user in members:
				members.append(user)
			
				self._groups[group].Set(
					"org.semplicelinux.usersd.group",
					"Members",
					members
				)
	
	def get_uids_with_users(self):
		"""
		A variant of the self._users dictionary, with UIDs as keys.
		"""
		
		result = {}
		
		for user, obj in self._users.items():
			
			result[obj.uid] = obj
		
		return result

	@usersd.objects.BaseObject.outside_timeout(
		"org.semplicelinux.usersd.user",
		out_signature="a{i(sss)}",
		sender_keyword="sender",
		connection_keyword="connection"
	)
	def GetUsers(self, sender, connection):
		"""
		This method returns a dictionary containing every user's UID as keys,
		and the username with the full name as values.
		"""
		
		result = {}
					
		for user, obj in self._users.items():
			result[obj.uid] = (user, obj.fullname, obj.home)
		
		return result

	@usersd.objects.BaseObject.outside_timeout(
		"org.semplicelinux.usersd.user",
		in_signature="s",
		out_signature="s",
		sender_keyword="sender",
		connection_keyword="connection"
	)
	def LookupUser(self, user, sender, connection):
		"""
		This method returns the object path for the given user.
		"""
		
		if user in self._users: return self._users[user].path
		
		return None

	@usersd.objects.BaseObject.outside_timeout(
		"org.semplicelinux.usersd.user",
		in_signature="ss",
		out_signature="b",
		sender_keyword="sender",
		connection_keyword="connection"
	)
	def CreateUser(self, user, fullname, sender, connection):
		"""
		This method returns the object path for the given user.
		"""
		
		if not is_authorized(
			sender,
			connection,
			"org.semplicelinux.usersd.add-user",
			True # user interaction
		):
			raise Exception("Not authorized")
		
		if usersd.user.User.add(user, fullname):
			# User created successfully, we should refresh the user list
			self._generate_users()
	
	@usersd.objects.BaseObject.outside_timeout(
		"org.semplicelinux.usersd.user",
		in_signature="sas",
		sender_keyword="sender",
		connection_keyword="connection"
	)
	def ShowUserCreationUI(self, display, groups, sender, connection):
		"""
		This method shows the user interface that permits to create a new
		user.
		"""
		
		if not is_authorized(
			sender,
			connection,
			"org.semplicelinux.usersd.add-user",
			True # user interaction
		):
			raise Exception("Not authorized")
		
		usersd.user.User.add_graphically(sender, self, display, groups)
	
if __name__ == "__main__":
		
	DBusGMainLoop(set_as_default=True)
	clss = Usersd()
	
	# Ladies and gentlemen...
	MainLoop.run()


usersd/objects.py
# -*- coding: utf-8 -*-
#
# usersd - user management daemon
# Copyright (C) 2014  Eugenio "g7" Paolantonio
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
# Authors:
#    Eugenio "g7" Paolantonio <me@medesimo.eu>
#

from usersd.common import MainLoop, is_authorized, get_user

import dbus
import dbus.service

from gi.repository import GLib, Polkit

# FIXME:
# Properties are not introspectable.
# Python DBus does not support properties OOTB unfortunately, so we end up
# with implementing the spec's Get(), GetAll(), Set() and PropertiesChanged()
# ourselves [1].
#
# While this works, having them introspectable would be a great thing.
# This of course needs some more work [2], and frankly I don't really care
# if those damn properties are introspectable or not! :D
#
# [1] http://stackoverflow.com/questions/3740903/python-dbus-how-to-export-interface-property
# [2] http://bazaar.launchpad.net/~laney/python-dbusmock/introspection-properties/view/head:/dbusmock/mockobject.py

class BaseObject(dbus.service.Object):
	"""
	A base object!
	"""
	
	# The DBus object path
	path = "/org/semplicelinux/usersd/baseobject"
	
	# The DBus interface name
	interface_name = "org.semplicelinux.usersd.baseobject"
	
	# The Polkit policy to use (None to disable authentication)
	polkit_policy = None
	
	# The properties to export to the Bus
	export_properties = []
	
	# The list of UIDs that can set properties without authentication.
	# NOTE: Please populate it from the __init__ method of your object!
	set_privileges = []

	def outside_timeout(*args, **kwargs):
		"""
		Decorator that ensures that the timeout doesn't elapse in the
		middle of our work.
		
		This mess of nested functions is needed because is not possible
		to use another decorator with @dbus.service.method. [1]
		We are then manually wrapping our decorator to python3-dbus's.
		
		[1] https://www.libreoffice.org/bugzilla/show_bug.cgi?id=22409
		"""
				
		def my_shiny_decorator(func):
		
			# Wrap the function around dbus.service.method
			func = dbus.service.method(*args, **kwargs)(func)
			
			def wrapper(self, *args, **kwargs):
													
				MainLoop.remove_timeout()
				result = func(self, *args, **kwargs)
				MainLoop.add_timeout()
				
				return result
			
			# Merge metadata, otherwise the method would not be
			# introspected
			wrapper.__name__ = func.__name__
			wrapper.__dict__.update(func.__dict__)
			wrapper.__module__ = wrapper.__module__
			
			return wrapper
		
		return my_shiny_decorator
	

	def __init__(self, bus_name):
		"""
		Initializes the object.
		"""
		
		super().__init__(bus_name, self.path)
	
	def store_property(self, name, value):
		"""
		Override this method to do things when the user changes
		a property.
		"""
		
		pass
	
	@outside_timeout(
		dbus_interface=dbus.PROPERTIES_IFACE,
		in_signature="ss",
		out_signature="v"
	)
	def Get(self, interface_name, property_name):
		"""
		An implementation of the Get() method of the
		properties interface.
		"""
		
		return self.GetAll(interface_name)[property_name]
	
	@outside_timeout(
		dbus_interface=dbus.PROPERTIES_IFACE,
		in_signature="s",
		out_signature="a{sv}"
	)
	def GetAll(self, interface_name):
		"""
		An implementation of the GetAll() method of the
		properties interface.
		"""
		
		if interface_name == self.interface_name:
			result = {}
			
			for prop in self.export_properties:
				try:
					result[prop.capitalize()] = getattr(self, prop)
				except:
					pass
			
			return result
		else:
			raise Exception(
				"org.semplicelinux.usersd.UnknownInterface",
				"The object does not implement the %s interface" % interface_name
			)
	
	@outside_timeout(
		dbus_interface=dbus.PROPERTIES_IFACE,
		in_signature="ssv",
		sender_keyword="sender",
		connection_keyword="connection"
	)
	def Set(self, interface_name, property_name, new_value, sender, connection):
		"""
		An implementation of the Set() method of the
		properties interface.
		"""
		
		if not get_user(sender) in self.set_privileges and (self.polkit_policy and not is_authorized(
			sender,
			connection,
			self.polkit_policy,
			True # user interaction
		)):
			raise Exception("E: Not authorized")
		
		self.store_property(property_name, new_value)


usersd/user.py
# -*- coding: utf-8 -*-
#
# usersd - user management daemon
# Copyright (C) 2014  Eugenio "g7" Paolantonio
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#
# Authors:
#    Eugenio "g7" Paolantonio <me@medesimo.eu>
#

import usersd.objects
import subprocess

from usersd.common import is_authorized, get_user

from usersd.common import Gtk, usersd_ui

from passlib.context import CryptContext

pwd_context = CryptContext(schemes=[
	"md5_crypt",
	"bcrypt",
	"sha1_crypt",
	"sun_md5_crypt",
	"sha256_crypt",
	"sha512_crypt",
])

default_encryption = "sha512_crypt"

MIN_PASSWORD_LENGTH = 4
USERNAME_ALLOWED_CHARS = "abcdefghijklmnopqrstuvwxyz0123456789.-"

class User(usersd.objects.BaseObject):
	"""
	The User object
	"""
	
	interface_name = "org.semplicelinux.usersd.user"
	export_properties = [
		"user",
		"uid",
		"gid",
		"fullname",
		"address",
		"phone",
		"other",
		"home",
		"shell"
	]
	polkit_policy = "org.semplicelinux.usersd.modify-user"
	
	@staticmethod
	def add(user, fullname, shell="/bin/bash"):
		"""
		This static method adds a new user, using the useradd command.
		The arguments are self-explanatory.
		
		NOTE: the password will *NOT* encrypted, so you have to encrypt
		it *BEFORE* calling this method.
		
		Returns True if the user has been created successfully, False
		if not.
		"""
		
		if subprocess.call((
			"/usr/sbin/useradd",
			user,
			"-m", # Create home directory
			"-U", # Create user group
			"-c", "%s,,,," % fullname, # Fullname
			"-s", shell, # Shell
		)) == 0:
			return True
		else:
			return False
	
	@staticmethod
	def add_graphically(sender, service, display, groups=[]):
		"""
		This static method allows the end-user to add a new user account.
		It differs from the add() method because this one uses a GTK+ Dialog
		to prompt the user for the required fields.
		
		It's useful for distributors who need to implement a graphical user
		management tool.
		"""
		
		# Create add_user_dialog
		usersd_ui._initialize(display, get_user(sender), service.get_uids_with_users())
		add_user_dialog = usersd_ui.AddUserDialog()
		
		# Connect response
		add_user_dialog.connect("response", User.on_add_user_dialog_response, add_user_dialog, service, groups)
		
		add_user_dialog.show()
	
	@staticmethod
	def on_add_user_dialog_response(dialog, response, parent, service, groups):
		"""
		Fired when a button on the add_user_dialog has been clicked.
		"""
		
		if response == Gtk.ResponseType.OK:
			# Ensure nothing is empty
			for obj in ("fullname", "username", "password", "confirm_password"):
				if parent.objects[obj].get_text() == "":
					parent.show_error(_("Please complete the entire form."))
					return False

			username = parent.objects.username.get_text()
			
			# Verify that the specified username is unique
			if username in service._users:
				parent.show_error(_("The username '%s' is already taken.") % username)
				return False
			
			# Verify that the specified username doesn't contain unallowed chars
			unallowed = []
			for char in username:
				if char not in USERNAME_ALLOWED_CHARS and char not in unallowed:
					unallowed.append(char)
			if unallowed:
				parent.show_error(_("The username must not contain the following characters: %s") % unallowed)
				return False
			
			# Verify passwords
			if not parent.objects.password.get_text() == parent.objects.confirm_password.get_text():
				parent.show_error(_("The passwords do not match."))
				return False
			
			# Check password length
			if not len(parent.objects.password.get_text()) >= MIN_PASSWORD_LENGTH:
				parent.show_error(_("The password should be of at least %s characters.") % MIN_PASSWORD_LENGTH)
				return False
			
			parent.hide_error()
						
			# Add the user
			if not User.add(username, parent.objects.fullname.get_text()):
				parent.show_error(_("Something went wrong while creating the new user."))
				return False
			
			# Add the user to the specified default groups
			service.AddGroupsToUser(username, groups)
			
			# Refresh
			service._generate_users()
			
			# Lookup for the newly created user
			if not username in service._users:
				parent.show_error(_("Something went wrong while creating the new user."))
				return False
			
			# Set password
			service._users[username].change_password(parent.objects.password.get_text())
			
		# Destroy the window
		dialog.destroy()
	
	@usersd.objects.BaseObject.outside_timeout(
		"org.semplicelinux.usersd.user",
		in_signature="b",
		out_signature="b",
		sender_keyword="sender",
		connection_keyword="connection"
	)
	def DeleteUser(self, with_home, sender, connection):
		"""
		Deletes the user.
		If with_home is True, the user's home directory will be deleted as well.
		
		Returns True if the user has been deleted successfully, False if not.
		"""
		
		if get_user(sender) == self.uid:
			# The sender can't remove itself!
			raise Exception("The sender can't remove itself!")
		
		if self.polkit_policy and not is_authorized(
			sender,
			connection,
			self.polkit_policy,
			True # user interaction
		):
			raise Exception("Not authorized")
		
		deluser_call = ["/usr/sbin/deluser", self.user]
		
		# deluser is picky with blank arguments, so we can't put an ""
		# in the place of --remove-home
		if with_home:
			deluser_call.append("--remove-home")
		
		if subprocess.call(deluser_call) == 0:
			self.service.remove_from_user_list(self.user)
			return True
		else:
			return False

	@usersd.objects.BaseObject.outside_timeout(
		"org.semplicelinux.usersd.user",
		in_signature="s",
		sender_keyword="sender",
		connection_keyword="connection"
	)
	def ChangePassword(self, display, sender, connection):
		"""
		This method returns the object path for the given user.
		"""
		
		if not get_user(sender) in self.set_privileges and (self.polkit_policy and not is_authorized(
			sender,
			connection,
			self.polkit_policy,
			True # user interaction
		)):
			raise Exception("Not authorized")
		
		# Create changepassword dialog
		usersd_ui._initialize(display, get_user(sender), self.service.get_uids_with_users())
		change_password_dialog = usersd_ui.ChangePasswordDialog(self.is_locked())
		
		# Connect response
		change_password_dialog.connect("response", self.on_change_password_dialog_response, change_password_dialog)
		
		change_password_dialog.show()

	def on_change_password_dialog_response(self, dialog, response, parent):
		"""
		Fired when a button on the change_password_dialog has been clicked.
		"""
		
		if response == Gtk.ResponseType.OK:
			# Verify old password
			if not parent.locked and not self.verify_password(parent.objects.old_password.get_text()):
				parent.show_error(_("Current password is not correct."))
				return False
			
			# Verify new passwords
			if not parent.objects.new_password.get_text() == parent.objects.confirm_new_password.get_text():
				parent.show_error(_("The new passwords do not match."))
				return False
			
			# Check password length
			if not len(parent.objects.new_password.get_text()) >= MIN_PASSWORD_LENGTH:
				parent.show_error(_("The new password should be of at least %s characters.") % MIN_PASSWORD_LENGTH)
				return False
			
			parent.hide_error()
			
			# Finally set password
			self.change_password(parent.objects.new_password.get_text())
			
		# Destroy the window
		dialog.destroy()
	
	def is_locked(self):
		"""
		Returns True if the user is locked (i.e. it doesn't have a password),
		False otherwise.
		"""
		
		status = True
		with open("/etc/shadow", "r") as f:
			for line in f:
				splt = line.split(":")
				if splt[0] == self.user:
					status = (splt[1] == "!")
					break
		
		splt = None
		return status

	def verify_password(self, oldpassword):
		"""
		Returns True if the password is verified, False otherwise.
		"""
		
		status = False
		with open("/etc/shadow", "r") as f:
			for line in f:
				splt = line.split(":")
				if splt[0] == self.user:
					status = pwd_context.verify(oldpassword, splt[1])
					break
		
		splt = None
		return status
	
	def change_password(self, newpassword):
		"""
		Changes the password.
		"""
		
		with open("/etc/shadow", "r") as f:
			lines = f.readlines()
		
		with open("/etc/shadow", "w") as f:
			for line in lines:
				splt = line.split(":")
				if splt[0] == self.user:
					# Change
					splt[1] = pwd_context.encrypt(newpassword, scheme=default_encryption)
				
				f.write(":".join(splt))
	
	def __init__(self, service, bus_name, passwd_entry):
		"""
		Initializes the object.
		"""
		
		self.service = service
		
		self.user, self.password, uid, gid, infos, self.home, self.shell = passwd_entry.split(":")
		
		# Parse GECOS
		infos = infos.split(",")
		self.fullname = self.address = self.phone = self.other = ""
		if len(infos) >= 1:
			self.fullname = infos[0]
		if len(infos) >= 2:
			self.address = infos[1]
		if len(infos) >= 3:
			self.phone = infos[2]
		if len(infos) >= 4:
			self.other = ",".join(infos[3:])
		
		self.uid = int(uid)
		self.gid = int(gid)
		
		# Ensure the actual user can write its own properties without authenticating
		self.set_privileges = [0, self.uid]
		
		self.path = "/org/semplicelinux/usersd/user/%s" % uid
		super().__init__(bus_name)
	
	def store_property(self, name, value):
		"""
		Stores the modified property in the /etc/passwd file.
		"""

		# Set value
		setattr(self, name[0].lower() + name[1:], value)
		
		# Save
		with open("/etc/passwd", "r") as f:
			# w+ doesn't seem to work lately...
			lines = f.readlines()
		
		with open("/etc/passwd", "w") as f:
			for line in lines:
				if line.split(":")[0] == self.user:
					# That's us!
					line = ":".join((
						self.user,
						self.password,
						str(self.uid),
						str(self.gid),
						",".join((
							self.fullname,
							self.address,
							self.phone,
							self.other
						)),
						self.home,
						self.shell
					)) + "\n"
				
				f.write(line)

